primitives/src/unique_vec.rs

```diff
@@ -0,0 +1,363 @@
+//! Module with [`UniqueVec`] type and related functional.
+
+#[cfg(not(feature = "std"))]
+use alloc::{borrow::ToOwned as _, format, string::String, vec::Vec};
+use core::{
+    borrow::Borrow,
+    fmt::{Debug, Display},
+    ops::Deref,
+};
+
+use iroha_schema::IntoSchema;
+use parity_scale_codec::{Decode, Encode};
+use serde::{Deserialize, Serialize};
+
+/// Creates a [`UniqueVec`](unique_vec::UniqueVec) from a list of values.
+///
+/// Works like [`vec!`] macro, but does not accept syntax for repeated values
+/// and might return [`Result`].
+#[macro_export]
+macro_rules! unique_vec {
+    () => {
+        $crate::unique_vec::UniqueVec::new()
+    };
+    ($($x:expr),+ $(,)?) => {{
+        let mut v = $crate::unique_vec::UniqueVec::new();
+        $(v.push($x);)+
+        v
+    }};
+}
+
+/// Wrapper type for [`Vec`] which ensures that all elements are unique.
+#[derive(
+    Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Deserialize, Serialize, Encode, Decode, IntoSchema,
+)]
+pub struct UniqueVec<T: PartialEq>(Vec<T>);
+
+impl<T: PartialEq> Default for UniqueVec<T> {
+    fn default() -> Self {
+        Self(Vec::default())
+    }
+}
+
+impl<T: PartialEq> UniqueVec<T> {
+    /// Create new [`UniqueVec`].
+    #[must_use]
+    pub const fn new() -> Self {
```

Comment:
if I'm not mistaken you can use `derive_more::{Constructor, Deref, AsRef}`

Reply:
`Deref` ans `AsRef` work fine, but `Contructor` generates `new(vec: Vec<T>) -> Self`, which is not what I need